scripts/user_retirement/docs/decisions/0001-extensibility.rst

```diff
@@ -0,0 +1,82 @@
+
+    .. code-block:: python3
+
+        - ['STEP_1', 'STEP_1_COMPLETE', 'DRIVER_NAME', 'method_name']
+        - ['STEP_2', 'STEP_2_COMPLETE', 'DRIVER_NAME', 'another_method_name']
+        # Example:
+        - ['RETIRING_EMAIL_LISTS', 'EMAIL_LISTS_COMPLETE', 'LMS', 'retirement_retire_mailings']
+        - ['RETIRING_ENROLLMENTS', 'ENROLLMENTS_COMPLETE', 'LMS', 'retirement_unenroll']
+
+#. Any logical step than can fail (ex: API call to a service that changes state) must be a separate method. This allows the retirement pipeline to handle failures gracefully, achieve greater idempotency, and enables retrying each individual step if necessary.
+
+#. As stated above, the entire configuration dictionary will be passed to the driver's `get_instance` method. This configuration is a dictionary version of the `configuration file`_ passed in to the retirement worker, and must contain all the configuration options that the driver needs to operate. The configuration dictionary will be passed to the driver's methods as the only argument. While scoping the configuration passed in to each driver would provide more secure isolation of secrets between drivers, it would also make it more difficult to share configuration (for instance in the case where a driver needs to access LMS or other configured service APIs as part of it's work)..
+
+#. The existing `utils/third_party_apis/*Api` objects will be removed from edx-platform as they are highly specific to edx.org and can be trivially reimplemented as one or more custom retirement drivers.
+
+Consequences
+============
+
+* Organizations will be able to extend the retirement pipeline without forking the codebase.
+* Organizations will be able to share retirement pipeline drivers with the community.
+* Organization-specific retirement steps will be removed from edx-platform.
+* Organizations using the existing third-party APIs will need to reimplement them as custom retirement drivers, or find a community-supported replacement. They will also need to update their deployment and configuration to use the new driver.
+
+Rejected Alternatives
+=====================
+
+Scoping driver configuration
+----------------------------
+Refactoring the retirement configuration to be more secure by scoping the configuration passed in to each driver was considered. This would provide better isolation of secrets between drivers and the default services, but would also make it more difficult to share configuration. For instance some drivers need to access LMS APIs, and doing so would require copying the LMS URL, client id, and client secret for each driver that needed them.
```

Comment:
From the perspective of an operator, if I need to pass all my secrets to a pipeline driver that someone else in the community wrote, I would be more reticent to use it. One response to this is that any retirement code you're running you should understand. However, this seems like it goes against the principal of least privilege.  If you still think this is the best idea, can you say some more about what kind of errors could happen and mention that we're choosing to go intentionally go against the principal of least privilege here.

Reply:
Thanks for testing my assumptions here, the initial goal was to reduce the impact of these changes (no config changes would have been necessary). Thinking about it more, I can see situations where you want different credentials for the same service as well. I don't think there's a way to change that without forcing a new configuration file format, but that's a lower priority than security.